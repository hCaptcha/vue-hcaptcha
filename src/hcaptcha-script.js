/**
 * Used to prevent loading api.js multiple times
 * @type {string}
 */
export const SCRIPT_ID = 'hcaptcha-api-script-id';

const HCAPTCHA_LOAD_FN_NAME = '_hcaptchaOnLoad';

let resolveFn;
let rejectFn;
const promise = new Promise((resolve, reject) => {
    resolveFn = resolve;
    rejectFn = reject;
});

/**
 * Async hcaptcha api.js loader.
 *
 * It makes sure `apiEndpoint` is loaded ONCE on the page despite calling this multiple times.
 *
 * Usage:
 * 1. import hcaptchaScript from './hcaptcha-script';
 * 2. when web component is mounted do:
 *      loadApiEndpointIfNotAlready('apiEndpoint', ...)
 *        .then(() => console.log('hcaptcha is loaded so it is safe to be used'))
 *        .catch((err) => console.error('failed to load the hcaptcha', err));
 *
 * @param config
 * @returns {Promise<void>}
 */
export function loadApiEndpointIfNotAlready(config) {
    if (window.hcaptcha) {
        // api.js is already present
        resolveFn();
        return promise;
    }
    if (document.getElementById(SCRIPT_ID)) {
        // api.js was already requested
        return promise;
    }
    // request api.js once
    window[HCAPTCHA_LOAD_FN_NAME] = resolveFn;
    const scriptSrc = getScriptSrc(config);
    const script = document.createElement('script');
    script.id = SCRIPT_ID;
    script.src = scriptSrc;
    script.async = true;
    script.defer = true;
    script.onerror = (event) => {
        // eslint-disable-next-line no-console
        console.error('Failed to load api: ' + scriptSrc, event);
        rejectFn('Failed to load api.js');
    };
    document.head.appendChild(script);
    return promise;
}

export function getScriptSrc(config) {
    let scriptSrc = config.apiEndpoint;
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'render', 'explicit');
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'onload', HCAPTCHA_LOAD_FN_NAME);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'recaptchacompat', config.reCaptchaCompat === false ? 'off' : null);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'hl', config.language);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'sentry', config.sentry);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'endpoint', config.endpoint);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'assethost', config.assethost);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'imghost', config.imghost);
    scriptSrc = addQueryParamIfDefined(scriptSrc, 'reportapi', config.reportapi);
    return scriptSrc;
}

export function addQueryParamIfDefined(url, queryName, queryValue) {
    if (queryValue !== undefined && queryValue !== null) {
        const link = url.includes('?') ? '&' : '?';
        return url + link + queryName + '=' + encodeURIComponent(queryValue);
    }
    return url;
}
