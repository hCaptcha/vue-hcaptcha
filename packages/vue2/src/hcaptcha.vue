<template>
  <div id="hcap-script" />
</template>

<script>
import {loadApiEndpointIfNotAlready} from './hcaptcha-script';

export default {
    name: 'VueHcaptcha',
    props: {
        sitekey: {
            type: String,
            required: true
        },
        theme: {
            type: String,
            default: undefined
        },
        size: {
            type: String,
            default: undefined
        },
        tabindex: {
            type: String,
            default: undefined
        },
        language: {
            type: String,
            default: undefined
        },
        reCaptchaCompat: {
            type: Boolean,
            default: true
        },
        challengeContainer: {
            type: String,
            default: undefined
        },
        rqdata: {
            type: String,
            default: undefined
        },
        sentry: {
            type: Boolean,
            default: true
        },
        apiEndpoint: {
            type: String,
            default: 'https://hcaptcha.com/1/api.js'
        },
        endpoint: {
            type: String,
            default: undefined
        },
        reportapi: {
            type: String,
            default: undefined
        },
        assethost: {
            type: String,
            default: undefined
        },
        imghost: {
            type: String,
            default: undefined
        },
    },
    data: () => {
        return {
            widgetId: null,
            hcaptcha: null,
            renderedCb : null,
        };
    },
    mounted() {
        return loadApiEndpointIfNotAlready(this.$props).then(this.onApiLoaded).catch(this.onError);
    },
    unmounted() {
        this.teardown();
    },
    destroyed() {
        this.teardown();
    },
    methods: {
        teardown() {
            if (this.widgetId) {
                this.hcaptcha.reset(this.widgetId);
                this.hcaptcha.remove(this.widgetId);
            }
        },
        onApiLoaded() {
            this.hcaptcha = window.hcaptcha;
            const opt = {
                sitekey: this.sitekey,
                theme: this.theme,
                size: this.size,
                tabindex: this.tabindex,
                'callback': this.onVerify,
                'expired-callback': this.onExpired,
                'chalexpired-callback': this.onChallengeExpired,
                'error-callback': this.onError,
                'open-callback': this.onOpen,
                'close-callback': this.onClose
            };
            if (this.challengeContainer) {
                opt['challenge-container'] = this.challengeContainer;
            }
            this.widgetId = this.hcaptcha.render(this.$el, opt);
            if (this.rqdata) {
                this.hcaptcha.setData(this.widgetId, {rqdata: this.rqdata});
            }
            this.onRendered();
        },
        execute() {
            if (this.widgetId) {
                this.hcaptcha.execute(this.widgetId);
                this.onExecuted();
            } else {
                // execute after el is rendered
                // we use a custom cb since `$on` was removed in vue3
                this.renderedCb = () => {
                    this.renderedCb = null;
                    this.execute();
                };
            }
        },
        reset() {
            if (this.widgetId) {
                this.hcaptcha.reset(this.widgetId);
                this.onReset();
            } else {
                this.$emit('error', 'Element is not rendered yet and thus cannot reset it. Wait for `rendered` event to safely call reset.');
            }
        },
        onRendered() {
            this.$emit('rendered');
            this.renderedCb && this.renderedCb();
        },
        onExecuted() {
            this.$emit('executed');
        },
        onReset() {
            this.$emit('reset');
        },
        onError(e) {
            this.$emit('error', e);
            this.reset();
        },
        onVerify() {
            const token = this.hcaptcha.getResponse(this.widgetId);
            const eKey = this.hcaptcha.getRespKey(this.widgetId);
            this.$emit('verify', token, eKey);
        },
        onExpired() {
            this.$emit('expired');
        },
        onChallengeExpired() {
            // vue3 will transform this `camelCase` event name into `kebab-case`
            this.$emit('challengeExpired');
        },
        onOpen() {
            this.$emit('opened');
        },
        onClose() {
            this.$emit('closed');
        }
    }
};
</script>
