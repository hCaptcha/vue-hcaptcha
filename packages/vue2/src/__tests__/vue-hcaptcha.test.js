import VueHcaptcha from '@/main';
import {DEBUG_SITE_KEY, getTestWrapper} from './utils';
import {getMockedHcaptcha, MOCK_EKEY, MOCK_TOKEN, MOCK_WIDGET_ID} from './hcaptcha.mock';
import {loadApiEndpointIfNotAlready} from '@/hcaptcha-script';
import {nextTick} from 'vue';

/*global describe, test, expect, beforeEach, jest*/

jest.mock('@/hcaptcha-script', () => ({
    loadApiEndpointIfNotAlready: jest.fn(() => Promise.resolve())
}));

describe('VueHCaptcha', () => {
    let wrapper;

    beforeEach(() => {
        // Clean dom before each test
        document.getElementsByTagName('html')[0].innerHTML = '';
        window.hcaptcha = getMockedHcaptcha();
        wrapper = getTestWrapper();
    });

    test('has functions', () => {
        const methods = VueHcaptcha.methods;
        expect(typeof methods.execute).toBe('function');
        expect(typeof methods.reset).toBe('function');
    });

    test('component renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    test('emits verify with token and ekey', () => {
        expect(wrapper.emitted()['verify']).toBeFalsy()
        wrapper.vm.onVerify();
        expect(wrapper.emitted().verify).toBeTruthy();
        expect(wrapper.emitted().verify[0][0]).toBe(MOCK_TOKEN);
        expect(wrapper.emitted().verify[0][1]).toBe(MOCK_EKEY);
    });

    test('emits error and calls reset', () => {
        const error = 'invalid-input-response';
        wrapper.vm.onError(error);
        expect(wrapper.emitted().error).toBeTruthy();
        expect(wrapper.emitted().error[0][0]).toBe(error);
        expect(wrapper.vm.hcaptcha.reset.mock.calls.length).toBe(1);
    });

    test('emits events', () => {
        expect(wrapper.emitted()['opened']).toBeFalsy();
        wrapper.vm.onOpen();
        expect(wrapper.emitted()['opened']).toBeTruthy();

        expect(wrapper.emitted()['closed']).toBeFalsy();
        wrapper.vm.onClose();
        expect(wrapper.emitted()['closed']).toBeTruthy();

        expect(wrapper.emitted()['expired']).toBeFalsy();
        wrapper.vm.onExpired();
        expect(wrapper.emitted()['expired']).toBeTruthy();

        expect(wrapper.emitted()['challengeExpired']).toBeFalsy();
        wrapper.vm.onChallengeExpired();
        expect(wrapper.emitted()['challengeExpired']).toBeTruthy();
    });

    test('reset emits event', async() => {
        expect(wrapper.emitted()['reset']).toBeFalsy();
        expect(wrapper.vm.hcaptcha.reset.mock.calls.length).toBe(0);
        wrapper.vm.reset();
        expect(wrapper.vm.hcaptcha.reset.mock.calls.length).toBe(1);
        expect(wrapper.emitted()['reset']).toBeTruthy();
    });

    test('execute emits event', async() => {
        expect(wrapper.emitted()['executed']).toBeFalsy();
        expect(wrapper.vm.hcaptcha.execute.mock.calls.length).toBe(0);
        wrapper.vm.execute();
        expect(wrapper.vm.hcaptcha.execute.mock.calls.length).toBe(1);
        expect(wrapper.emitted()['executed']).toBeTruthy();
    });

    test('calling execute before rendering', async() => {
        wrapper = getTestWrapper();
        wrapper.vm.execute();
        expect(window.hcaptcha.execute.mock.calls.length).toBe(0);
        await nextTick();
        expect(window.hcaptcha.execute.mock.calls.length).toBe(1);
    });

    test('el renders after api loads and a widget id is set', async() => {
        const props = {
            sitekey: DEBUG_SITE_KEY,
            theme: 'dark',
            size: 'compact',
            tabindex: '10',
            language: 'ro',
            reCaptchaCompat: true,
            challengeContainer: 'challenge-container-id',
            rqdata: null,
            sentry: false,
            apiEndpoint: 'https://example.com/api.js',
            endpoint: 'https://endpoint.example.com',
            reportapi: 'https://reportapi.example.com',
            assethost: 'https://assethost.example.com',
            imghost: 'https://imghost.example.com',
        };
        window.hcaptcha = getMockedHcaptcha();
        wrapper = getTestWrapper(props);
        await nextTick();
        expect(wrapper.vm.$data.widgetId).toBe(MOCK_WIDGET_ID);
        expect(wrapper.vm.hcaptcha.render.mock.calls.length).toBe(1);
        expect(wrapper.vm.hcaptcha.render.mock.calls[0][1]).toMatchObject({
            sitekey: props.sitekey,
            theme: props.theme,
            size: props.size,
            tabindex: props.tabindex
        });
    });

    test('mounted - loads api endpoint', async() => {
        const nrCalls = loadApiEndpointIfNotAlready.mock.calls.length;
        wrapper = getTestWrapper();
        expect(loadApiEndpointIfNotAlready.mock.calls.length).toBe(nrCalls + 1);
    });

    test('rqdata is set upon rendering', async() => {
        const props = {
            sitekey: DEBUG_SITE_KEY,
            rqdata: 'mock-rqdata'
        };
        window.hcaptcha = getMockedHcaptcha();
        wrapper = getTestWrapper(props);
        await nextTick();
        expect(wrapper.vm.$data.widgetId).toBe(MOCK_WIDGET_ID);
        expect(wrapper.vm.hcaptcha.render.mock.calls.length).toBe(1);
        expect(wrapper.vm.hcaptcha.setData.mock.calls[0][1]).toMatchObject({
            rqdata: props.rqdata
        });
    });
});
