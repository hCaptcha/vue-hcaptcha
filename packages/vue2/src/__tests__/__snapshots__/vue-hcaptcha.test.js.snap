// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue2 hCaptcha component renders correctly 1`] = `
<div
  id="hcap-script"
/>
`;
